gensim/models/word2vec_inner.pyx

```diff
@@ -45,6 +54,144 @@ cdef REAL_t[EXP_TABLE_SIZE] LOG_TABLE
 cdef int ONE = 1
 cdef REAL_t ONEF = <REAL_t>1.0
 
+
+@cython.final
+cdef class CythonVocab:
+    def __init__(self, pyvocab, hs=0):
+        cdef VocabItem word
+
+        for py_token, vocab_item in iteritems(pyvocab):
+            token = any2utf8(py_token)
+            word.index = vocab_item.index
+            word.sample_int = vocab_item.sample_int
+
+            if hs:
+                word.code = <np.uint8_t *>np.PyArray_DATA(vocab_item.code)
+                word.code_len = <int>len(vocab_item.code)
+                word.point = <np.uint32_t *>np.PyArray_DATA(vocab_item.point)
+
+            self.vocab[token] = word
+
+    cdef cvocab_t* get_vocab_ptr(self) nogil except *:
+        return &self.vocab
+
+
+def rebuild_cython_line_sentence(source, max_sentence_length):
+    return CythonLineSentence(source, max_sentence_length=max_sentence_length)
+
+
+cdef bytes to_bytes(key):
+    if isinstance(key, bytes):
+        return <bytes>key
+    else:
+        return key.encode('utf8')
+
+
+@cython.final
+cdef class CythonLineSentence:
+    def __cinit__(self, source, offset=0, max_sentence_length=MAX_SENTENCE_LEN):
+        self._thisptr = new FastLineSentence(to_bytes(source), offset)
+
+    def __init__(self, source, offset=0, max_sentence_length=MAX_SENTENCE_LEN):
+        self.source = to_bytes(source)
+        self.offset = offset
+        self.max_sentence_length = max_sentence_length
+        self.max_words_in_batch = max_sentence_length
+
+    def __dealloc__(self):
+        if self._thisptr != NULL:
+            del self._thisptr
+
+    cpdef bool_t is_eof(self) nogil:
+        return self._thisptr.IsEof()
+
+    cpdef vector[string] read_sentence(self) nogil except *:
+        return self._thisptr.ReadSentence()
+
+    cpdef vector[vector[string]] _read_chunked_sentence(self) nogil except *:
+        cdef vector[string] sent = self.read_sentence()
+        return self._chunk_sentence(sent)
+
+    cpdef vector[vector[string]] _chunk_sentence(self, vector[string] sent) nogil:
+        cdef vector[vector[string]] res
+        cdef vector[string] chunk
+        cdef size_t cur_idx = 0
+
+        if sent.size() > self.max_sentence_length:
+            while cur_idx < sent.size():
+                chunk.clear()
+                for i in range(cur_idx, min(cur_idx + self.max_sentence_length, sent.size())):
+                    chunk.push_back(sent[i])
+
+                res.push_back(chunk)
+                cur_idx += chunk.size()
+        else:
+            res.push_back(sent)
+
+        return res
+
+    cpdef void reset(self) nogil:
+        self._thisptr.Reset()
+
+    def __iter__(self):
+        self.reset()
+        while not self.is_eof():
+            chunked_sentence = self._read_chunked_sentence()
+            for chunk in chunked_sentence:
+                if not chunk.empty():
+                    yield chunk
+
+    def __reduce__(self):
+        return rebuild_cython_line_sentence, (self.source, self.max_sentence_length)
```

Comment:
need clarification, I don't understand, why this needed

Reply:
This function is needed for `pickle` module to correctly serialize this class and transfer between processes. I added this function when we had multistream vocab building (via multiprocessing).

Reply:
thanks, deserve a comment in code (or docstring, up to you)